### PR TITLE
Changelings now get an option to be revived by defibrillator when faking death

### PR DESCRIPTION
**Pull request by Contrabang**
## What Does This PR Do

When a changeling is faking death, and they get defibrillated, they can now be revived like any normal dead person.

If the changeling accepts, it will appear like a normal defib happened, and the changeling loses all aheal/revive powers
If the changeling denies (doesn't click on the status) it will appear like the ghost isn't in the body
If its a non-changeling form of fake death, like capulettium, it will appear like it used to.

A white-box will appear around the status effect to indicate it has been accepted. It can be toggled for the 5-seconds while being defibbed, by default it is off.

Also adds more safeties to the cling Revive power to prevent it from being stacked or from being used while alive.

## Why It's Good For The Game

Its very awkward when a changeling is forced to revive or be cloned or just left in morgue because security is watching their body like a hawk because the defib failed for no clear reason. Helps against security meta-ing changelings.

## Images of changes

image

## Testing

2 clients, revived myself a bunch.
I want to do more testing before undrafting.

## Changelog
:cl:
add: Changelings now get an option to be revived by defibrillator when faking death
/:cl:

### CODE REVIEW

**Comment by Coolrune206** [comment]
> Helps against security meta-ing changelings.

Wouldn't really call this meta. The defibrillator failing because the creature is still very much alive fits the lore of the antagonist, though it is reasonable that they'd be able to adapt and pretend to come back to life with the defibrillator, simply mimicking the "ghost not in body" message feels an odd choice for the other outcome.

**Comment by tipoima** [comment]
> Wouldn't really call this meta. The defibrillator failing because the creature is still very much alive fits the lore of the antagonist
Technically, anyone would be considered "alive" in that situation. All you're doing is restarting a still living heart. 

> simply mimicking the "ghost not in body" message feels an odd choice for the other outcome
The player could also just leave the body to force that message instead. Might as well make it the default (and not have a message that specifically triggers for changelings)

**Comment by Contrabang** [comment]
> Wouldn't really call this meta. The defibrillator failing because the creature is still very much alive fits the lore of the antagonist, though it is reasonable that they'd be able to adapt and pretend to come back to life with the defibrillator, simply mimicking the "ghost not in body" message feels an odd choice for the other outcome.

i think its better to buff them in this stealth way, than having the changeling forced to get up and armblade everyone nearby in medbay, or be forced into cloning/the morgue. As for the "ghost not in body" message, the way its said in game is "Patient's brain is unresponsive. Further attempts may succeed." which I think still aptly fits the description of the changeling choosing to not be revived

File: code/datums/components/defibrillator.dm [hunk 1]
```diff
@@ -238,37 +243,39 @@
...
 		if(!ghost.can_reenter_corpse || target.suiciding) // DNR or AntagHUD
-			user.visible_message("<span class='boldnotice'>[defib_ref] buzzes: Resuscitation failed - No electrical brain activity detected.</span>")
+			failure_message = "<span class='boldnotice'>[defib_ref] buzzes: Resuscitation failed - No electrical brain activity detected.</span>"
+		else
+			failure_message = "<span class='boldnotice'>[defib_ref] buzzes: Resuscitation failed - Patient's brain is unresponsive. Further attempts may succeed.</span>"
+	else if(HAS_TRAIT(target, TRAIT_FAKEDEATH))
+		if(signal_result & COMPONENT_DEFIB_FAKEDEATH_DENIED)
+			failure_message = "<span class='boldnotice'>[defib_ref] buzzes: Resuscitation failed - Patient's brain is unresponsive. Further attempts may succeed.</span>"
+		else if(signal_result & COMPONENT_DEFIB_FAKEDEATH_ACCEPTED)
+			// as much as I hate that this is here, it has to come after the `Patient is not in a valid state. Operation aborted.` check.
```
Comment (DGamerL) [inline code/datums/components/defibrillator.dm:269]: This seems fine to be here? Am I missing something?

**Review by DGamerL — APPROVED** [review]
Looks pretty good as far as I can see

File: code/datums/components/defibrillator.dm (hunk 1 above)
Comment (Contrabang) [reply code/datums/components/defibrillator.dm:269]: Ideally there would only be a True and false condition, but theres 3 conditions here

File: code/_onclick/hud/alert.dm [hunk 2]
```diff
@@ -870,3 +870,56 @@ so as to remain in compliance with the most up-to-date laws."
...
+
+/atom/movable/screen/alert/changeling_defib_revive
+	name = "Your heart is being defibrillated."
+	desc = "Click this status to be revived from fake death."
+	icon_state = "template"
+	timeout = 10 SECONDS
+	var/do_we_revive = FALSE
+	var/image/toggle_overlay
+
+/atom/movable/screen/alert/changeling_defib_revive/Initialize(mapload, parent_unit,)
```
Comment (lewcc) [inline code/_onclick/hud/alert.dm:882]:
```suggestion
/atom/movable/screen/alert/changeling_defib_revive/Initialize(mapload, parent_unit)
```

File: code/_onclick/hud/alert.dm [hunk 3]
```diff
@@ -870,3 +870,56 @@ so as to remain in compliance with the most up-to-date laws."
...
+		overlays -= toggle_overlay
+
+/atom/movable/screen/alert/changeling_defib_revive/proc/on_defib_revive()
+	. = COMPONENT_DEFIB_FAKEDEATH_DENIED
+	if(do_we_revive)
+		var/datum/antagonist/changeling/cling = owner.mind.has_antag_datum(/datum/antagonist/changeling)
+		cling.remove_specific_power(/datum/action/changeling/revive)
+		cling.regenerating = FALSE
+		var/heart_name = "heart"
+		if(iscarbon(owner))
```
Comment (lewcc) [inline code/_onclick/hud/alert.dm:919]: Is there any chance that the owner would not be a carbon here?

File: code/_onclick/hud/alert.dm [hunk 4]
```diff
@@ -870,3 +870,56 @@ so as to remain in compliance with the most up-to-date laws."
...
+/atom/movable/screen/alert/changeling_defib_revive/Initialize(mapload, parent_unit,)
+	. = ..()
+	var/image/defib_appearance = image(parent_unit)
+	defib_appearance.layer = FLOAT_LAYER
+	defib_appearance.plane = FLOAT_PLANE
+	overlays += defib_appearance
+
+/atom/movable/screen/alert/changeling_defib_revive/attach_owner(mob/new_owner)
+	. = ..()
+	RegisterSignal(owner, COMSIG_LIVING_DEFIBBED, PROC_REF(on_defib_revive))
```
Comment (lewcc) [inline code/_onclick/hud/alert.dm:891]: Do you know if you might need to remove this signal registration on detach from owner/destroy?

**Review by lewcc — CHANGES_REQUESTED** [review]
Just some small things, mostly looks good

File: code/_onclick/hud/alert.dm (hunk 3 above)
Comment (Contrabang) [reply code/_onclick/hud/alert.dm:919]: I don't think so, but I left it like that to be safe

File: code/_onclick/hud/alert.dm (hunk 4 above)
Comment (Contrabang) [reply code/_onclick/hud/alert.dm:891]: Don't think its needed, but I'll add it.

**Review by lewcc — APPROVED** [review]

**Pull request merged**